File: skills/video-production/scripts/delivery_qc.mjs
#!/usr/bin/env node
/* Check a finished file against a platform delivery spec.
 *
 *   node delivery_qc.mjs --video out.mp4 --spec youtube-16x9 [--json qc.json]
 *   node delivery_qc.mjs --video out.mp4 --spec-file custom.json
 *   node delivery_qc.mjs --list
 *
 * Answers the question `hyperframes check` cannot: the composition is valid, but is
 * the FILE acceptable where it is going? Container, codec, resolution, frame rate,
 * duration bounds, loudness, true peak, silence, bitrate.
 *
 * Specs live in assets/delivery-specs.json. They are derived from published
 * platform guidance where it exists; the ones marked "unverified" there are
 * community consensus and should be confirmed before a contractual delivery.
 */
import { spawnSync } from "node:child_process";
import { readFileSync, writeFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const HERE = dirname(fileURLToPath(import.meta.url));
const SPECS = resolve(HERE, "..", "assets", "delivery-specs.json");

const run = (cmd, args) =>
  spawnSync(cmd, args, { encoding: "utf8", maxBuffer: 64 * 1024 * 1024 });

function probe(video) {
  const r = run("ffprobe", ["-v", "error", "-show_format", "-show_streams",
                            "-of", "json", video]);
  if (r.status !== 0) {
    console.error("ffprobe failed: " + (r.stderr || "").trim());
    process.exit(1);
  }
  return JSON.parse(r.stdout);
}

function loudness(video) {
  const r = run("ffmpeg", ["-hide_banner", "-nostats", "-i", video,
                           "-af", "ebur128=peak=true", "-f", "null", "-"]);
  const out = r.stderr || "";
  const tail = out.includes("Summary:") ? out.slice(out.lastIndexOf("Summary:")) : out;
  const got = {};
  for (const line of tail.split("\n")) {
    const s = line.trim();
    const parts = s.split(/\s+/);
    if (s.startsWith("I:") && s.includes("LUFS")) got.lufs = parseFloat(parts[1]);
    else if (s.startsWith("Peak:") && s.includes("dBFS")) got.true_peak = parseFloat(parts[1]);
    else if (s.startsWith("LRA:") && s.includes("LU") && got.lra === undefined)
      got.lra = parseFloat(parts[1]);
  }
  return got;
}

function silence(video, floor = -50, dur = 0.25) {
  const r = run("ffmpeg", ["-v", "error", "-i", video,
                           "-af", `silencedetect=n=${floor}dB:d=${dur}`, "-f", "null", "-"]);
  return (r.stderr || "").split("\n").map((l) => l.trim()).filter((l) => l.includes("silence_start"));
}

function integrity(video) {
  const r = run("ffmpeg", ["-v", "error", "-i", video, "-f", "null", "-"]);
  return (r.stderr || "").trim();
}

/* A dB ceiling and a ratio are float-valued fields, and an integral one still
 * reads as a measurement: "-1.0 dBFS", not "-1". `lufs` is written as a whole
 * number in the spec file and stays one. */
const pyNum = (v) => (Number.isInteger(v) ? v.toFixed(1) : String(v));

/* 1080/1920 is exactly 0.5625 — a true tie, and the aspect tolerance is only
 * ±0.01, so which way it breaks is a verdict, not a decoration. Round half to
 * even, the same rule the rest of the toolchain uses. */
const pyRound = (v, d) => {
  const m = 10 ** d, x = v * m;
  if (Math.abs(x - Math.trunc(x)) === 0.5) {
    const t = Math.trunc(x);
    return (t % 2 === 0 ? t : t + Math.sign(x)) / m;
  }
  return Math.round(x) / m;
};
const pyList = (a) => "[" + a.map((x) => (typeof x === "string" ? `'${x}'` : x)).join(", ") + "]";
const pad = (s, w) => s + " ".repeat(Math.max(0, w - s.length));

class Report {
  constructor() { this.rows = []; }
  add(name, ok, detail, warn = false) { this.rows.push({ name, ok, detail, warn }); }
  render() {
    const w = Math.max(...this.rows.map((r) => r.name.length)) + 2;
    const bar = "=".repeat(78);
    console.log(bar);
    console.log(pad("DELIVERY QC", w) + "RESULT   DETAIL");
    console.log(bar);
    let bad = 0;
    for (const r of this.rows) {
      const tag = r.ok ? " ok  " : (r.warn ? "warn " : "FAIL ");
      if (!r.ok && !r.warn) bad++;
      console.log(pad(r.name, w) + tag + "   " + r.detail);
    }
    console.log(bar);
    console.log(bad === 0 ? "PASSED" : `${bad} CHECK(S) FAILED`);
    return bad;
  }
}

const near = (a, b, tol) => Math.abs(a - b) <= tol;

// ── main ────────────────────────────────────────────────────────────────────

const argv = process.argv.slice(2);
const opt = (n) => { const i = argv.indexOf("--" + n); return i === -1 ? undefined : argv[i + 1]; };
const flag = (n) => argv.includes("--" + n);
const die = (m) => { console.error(m); process.exit(2); };

const catalog = JSON.parse(readFileSync(SPECS, "utf8"));

if (flag("list")) {
  for (const [k, v] of Object.entries(catalog)) {
    if (k.startsWith("_")) continue;
    console.log(pad(k, 22) + " " + (v.label || ""));
  }
  process.exit(0);
}

const video = opt("video");
if (!video) die("--video is required");

let spec;
if (opt("spec-file")) spec = JSON.parse(readFileSync(opt("spec-file"), "utf8"));
else if (opt("spec")) {
  if (!(opt("spec") in catalog)) die(`unknown spec '${opt("spec")}' — try --list`);
  spec = catalog[opt("spec")];
} else die("pass --spec or --spec-file");

const info = probe(video);
const fmt = info.format || {};
const vs = info.streams.find((s) => s.codec_type === "video");
const as_ = info.streams.find((s) => s.codec_type === "audio");

const rep = new Report();

const err = integrity(video);
rep.add("file integrity", !err, !err ? "clean" : err.slice(0, 70));

if (!vs) {
  rep.add("video stream", false, "none found");
  process.exit(rep.render() ? 1 : 0);
}

const w = parseInt(vs.width, 10), h = parseInt(vs.height, 10);
if (spec.resolution) {
  const ok = w === spec.resolution[0] && h === spec.resolution[1];
  rep.add("resolution", ok,
    `${w}x${h}` + (ok ? "" : ` (want ${spec.resolution[0]}x${spec.resolution[1]})`));
}

if (spec.aspect) {
  const got = pyRound(w / h, 3);
  const ok = near(got, spec.aspect, 0.01);
  rep.add("aspect ratio", ok,
          `${pyNum(got)}` + (ok ? "" : ` (want ${pyNum(spec.aspect)})`));
}

const [num, den] = (vs.r_frame_rate || "0/1").split("/");
const fps = parseFloat(num) / (parseFloat(den) || 1);
if (spec.fps) {
  const ok = spec.fps.some((x) => near(fps, x, 0.05));
  rep.add("frame rate", ok, fps.toFixed(2) + (ok ? "" : ` (want ${pyList(spec.fps)})`));
}

const dur = parseFloat(fmt.duration || 0);
const [lo, hi] = spec.duration_s || [null, null];
if (lo !== null || hi !== null) {
  const ok = (lo === null || dur >= lo - 0.05) && (hi === null || dur <= hi + 0.05);
  rep.add("duration", ok, `${dur.toFixed(2)}s (allowed ${lo}-${hi})`);
}

if (spec.video_codec) {
  const ok = spec.video_codec.includes(vs.codec_name);
  rep.add("video codec", ok, vs.codec_name + (ok ? "" : ` (want ${pyList(spec.video_codec)})`));
}

const bitrate = fmt.bit_rate ? parseInt(fmt.bit_rate, 10) / 1e6 : null;
if (spec.max_bitrate_mbps && bitrate) {
  rep.add("bitrate", bitrate <= spec.max_bitrate_mbps,
          `${bitrate.toFixed(1)} Mbps (max ${spec.max_bitrate_mbps})`);
}

const sizeMb = parseInt(fmt.size || 0, 10) / 1048576;
if (spec.max_size_mb) {
  rep.add("file size", sizeMb <= spec.max_size_mb,
          `${sizeMb.toFixed(1)} MB (max ${spec.max_size_mb})`);
}

if (!as_) {
  rep.add("audio stream", spec.audio_required === false,
          spec.audio_required === false ? "none — silent deliverable" : "missing");
} else {
  if (spec.audio_codec) rep.add("audio codec", spec.audio_codec.includes(as_.codec_name), as_.codec_name);
  if (spec.sample_rate)
    rep.add("sample rate", spec.sample_rate.includes(parseInt(as_.sample_rate || 0, 10)),
            `${as_.sample_rate} Hz`);

  const ln = loudness(video);
  if (spec.lufs !== undefined && spec.lufs !== null && ln.lufs !== undefined) {
    const tol = spec.lufs_tolerance ?? 1.0;
    rep.add("loudness", near(ln.lufs, spec.lufs, tol),
            `${pyNum(ln.lufs)} LUFS (target ${spec.lufs} ±${pyNum(tol)})`);
  }
  if (spec.true_peak !== undefined && spec.true_peak !== null && ln.true_peak !== undefined) {
    rep.add("true peak", ln.true_peak <= spec.true_peak + 0.05,
            `${pyNum(ln.true_peak)} dBFS (ceiling ${pyNum(spec.true_peak)})`);
  }

  if (spec.no_silence !== false) {
    const sil = silence(video);
    rep.add("no dropout", !sil.length, !sil.length ? "continuous" : `${sil.length} gap(s)`);
  }
}

if (spec.notes) rep.add("spec notes", true, spec.notes, true);

const bad = rep.render();

if (opt("json")) {
  writeFileSync(opt("json"), JSON.stringify({
    video,
    spec: opt("spec") || opt("spec-file"),
    passed: bad === 0,
    checks: rep.rows.map((r) => ({ name: r.name, ok: r.ok, detail: r.detail, warn: r.warn })),
  }, null, 2), "utf8");
}

process.exit(bad ? 1 : 0);

File: skills/video-production/scripts/delivery_qc.py
#!/usr/bin/env python3
"""Check a finished file against a platform delivery spec.

    python delivery_qc.py --video out.mp4 --spec youtube-16x9 [--json qc.json]
    python delivery_qc.py --video out.mp4 --spec-file custom.json
    python delivery_qc.py --list

Answers the question `hyperframes check` cannot: the composition is valid, but is
the FILE acceptable where it is going? Container, codec, resolution, frame rate,
duration bounds, loudness, true peak, silence, bitrate.

Specs live in assets/delivery-specs.json. They are derived from published
platform guidance where it exists; the ones marked "unverified" there are
community consensus and should be confirmed before a contractual delivery.
"""
import argparse
import json
import subprocess
import sys
from pathlib import Path

try:
    sys.stdout.reconfigure(encoding="utf-8")
    sys.stderr.reconfigure(encoding="utf-8")
except Exception:
    pass

SPECS = Path(__file__).resolve().parent.parent / "assets" / "delivery-specs.json"


def run(cmd):
    return subprocess.run(cmd, capture_output=True, text=True, encoding="utf-8",
                          errors="replace")


def probe(video):
    r = run(["ffprobe", "-v", "error", "-show_format", "-show_streams",
             "-of", "json", video])
    if r.returncode != 0:
        raise SystemExit("ffprobe failed: " + (r.stderr or "").strip())
    return json.loads(r.stdout)


def loudness(video):
    r = run(["ffmpeg", "-hide_banner", "-nostats", "-i", video,
             "-af", "ebur128=peak=true", "-f", "null", "-"])
    out, got = r.stderr or "", {}
    tail = out[out.rfind("Summary:"):] if "Summary:" in out else out
    for line in tail.splitlines():
        s = line.strip()
        if s.startswith("I:") and "LUFS" in s:
            got["lufs"] = float(s.split()[1])
        elif s.startswith("Peak:") and "dBFS" in s:
            got["true_peak"] = float(s.split()[1])
        elif s.startswith("LRA:") and "LU" in s and "lra" not in got:
            got["lra"] = float(s.split()[1])
    return got


def silence(video, floor=-50, dur=0.25):
    r = run(["ffmpeg", "-v", "error", "-i", video,
             "-af", f"silencedetect=n={floor}dB:d={dur}", "-f", "null", "-"])
    return [l.strip() for l in (r.stderr or "").splitlines() if "silence_start" in l]


def integrity(video):
    r = run(["ffmpeg", "-v", "error", "-i", video, "-f", "null", "-"])
    return (r.stderr or "").strip()


class Report:
    def __init__(self):
        self.rows = []

    def add(self, name, ok, detail, warn=False):
        self.rows.append((name, ok, detail, warn))

    def render(self):
        w = max(len(n) for n, *_ in self.rows) + 2
        print("=" * 78)
        print("DELIVERY QC".ljust(w) + "RESULT   DETAIL")
        print("=" * 78)
        bad = 0
        for name, ok, detail, warn in self.rows:
            tag = " ok  " if ok else ("warn " if warn else "FAIL ")
            if not ok and not warn:
                bad += 1
            print(f"{name.ljust(w)}{tag}   {detail}")
        print("=" * 78)
        print("PASSED" if bad == 0 else f"{bad} CHECK(S) FAILED")
        return bad


def near(a, b, tol):
    return abs(a - b) <= tol


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--video")
    ap.add_argument("--spec", help="a key from assets/delivery-specs.json")
    ap.add_argument("--spec-file", help="a JSON file holding one spec object")
    ap.add_argument("--json", help="write the result here")
    ap.add_argument("--list", action="store_true", help="list built-in specs")
    a = ap.parse_args()

    catalog = json.loads(SPECS.read_text(encoding="utf-8"))

    if a.list:
        for k, v in catalog.items():
            if k.startswith("_"):
                continue
            print(f"{k:<22} {v.get('label', '')}")
        return 0

    if not a.video:
        ap.error("--video is required")
    if a.spec_file:
        spec = json.loads(Path(a.spec_file).read_text(encoding="utf-8"))
    elif a.spec:
        if a.spec not in catalog:
            ap.error(f"unknown spec '{a.spec}' — try --list")
        spec = catalog[a.spec]
    else:
        ap.error("pass --spec or --spec-file")

    info = probe(a.video)
    fmt = info.get("format", {})
    vs = next((s for s in info["streams"] if s["codec_type"] == "video"), None)
    as_ = next((s for s in info["streams"] if s["codec_type"] == "audio"), None)

    rep = Report()

    err = integrity(a.video)
    rep.add("file integrity", not err, "clean" if not err else err[:70])

    if vs is None:
        rep.add("video stream", False, "none found")
        return rep.render()

    w, h = int(vs["width"]), int(vs["height"])
    want = spec.get("resolution")
    if want:
        ok = [w, h] == want
        rep.add("resolution", ok, f"{w}x{h}" + ("" if ok else f" (want {want[0]}x{want[1]})"))

    ar = spec.get("aspect")
    if ar:
        got = round(w / h, 3)
        ok = near(got, ar, 0.01)
        rep.add("aspect ratio", ok, f"{got}" + ("" if ok else f" (want {ar})"))

    num, den = (vs.get("r_frame_rate") or "0/1").split("/")
    fps = float(num) / float(den or 1)
    if spec.get("fps"):
        ok = any(near(fps, f, 0.05) for f in spec["fps"])
        rep.add("frame rate", ok, f"{fps:.2f}" + ("" if ok else f" (want {spec['fps']})"))

    dur = float(fmt.get("duration", 0))
    lo, hi = spec.get("duration_s", [None, None])
    if lo is not None or hi is not None:
        ok = (lo is None or dur >= lo - 0.05) and (hi is None or dur <= hi + 0.05)
        rep.add("duration", ok, f"{dur:.2f}s (allowed {lo}-{hi})")

    if spec.get("video_codec"):
        ok = vs["codec_name"] in spec["video_codec"]
        rep.add("video codec", ok, vs["codec_name"] + ("" if ok else f" (want {spec['video_codec']})"))

    bitrate = int(fmt.get("bit_rate", 0)) / 1_000_000 if fmt.get("bit_rate") else None
    if spec.get("max_bitrate_mbps") and bitrate:
        ok = bitrate <= spec["max_bitrate_mbps"]
        rep.add("bitrate", ok, f"{bitrate:.1f} Mbps (max {spec['max_bitrate_mbps']})")

    size_mb = int(fmt.get("size", 0)) / 1_048_576
    if spec.get("max_size_mb"):
        ok = size_mb <= spec["max_size_mb"]
        rep.add("file size", ok, f"{size_mb:.1f} MB (max {spec['max_size_mb']})")

    if as_ is None:
        rep.add("audio stream", spec.get("audio_required", True) is False,
                "none — silent deliverable" if spec.get("audio_required") is False else "missing")
    else:
        if spec.get("audio_codec"):
            ok = as_["codec_name"] in spec["audio_codec"]
            rep.add("audio codec", ok, as_["codec_name"])
        if spec.get("sample_rate"):
            ok = int(as_.get("sample_rate", 0)) in spec["sample_rate"]
            rep.add("sample rate", ok, f"{as_.get('sample_rate')} Hz")

        ln = loudness(a.video)
        if spec.get("lufs") is not None and "lufs" in ln:
            tol = spec.get("lufs_tolerance", 1.0)
            ok = near(ln["lufs"], spec["lufs"], tol)
            rep.add("loudness", ok, f"{ln['lufs']} LUFS (target {spec['lufs']} ±{tol})")
        if spec.get("true_peak") is not None and "true_peak" in ln:
            ok = ln["true_peak"] <= spec["true_peak"] + 0.05
            rep.add("true peak", ok, f"{ln['true_peak']} dBFS (ceiling {spec['true_peak']})")

        if spec.get("no_silence", True):
            sil = silence(a.video)
            rep.add("no dropout", not sil, "continuous" if not sil else f"{len(sil)} gap(s)")

    if spec.get("notes"):
        rep.add("spec notes", True, spec["notes"], warn=True)

    bad = rep.render()

    if a.json:
        Path(a.json).write_text(json.dumps({
            "video": a.video,
            "spec": a.spec or a.spec_file,
            "passed": bad == 0,
            "checks": [{"name": n, "ok": o, "detail": d, "warn": w}
                       for n, o, d, w in rep.rows],
        }, ensure_ascii=False, indent=2), encoding="utf-8")

    return 1 if bad else 0


if __name__ == "__main__":
    sys.exit(main())
